pymc3/gp/gp.py

```diff
@@ -104,13 +104,13 @@ def __init__(self, mean_func=Zero(), cov_func=Constant(0.0)):
 
     def _build_prior(self, name, X, reparameterize=True, **kwargs):
         mu = self.mean_func(X)
-        chol = cholesky(stabilize(self.cov_func(X)))
+        cov = stabilize(self.cov_func(X))
         shape = infer_shape(X, kwargs.pop("shape", None))
         if reparameterize:
             v = pm.Normal(name + "_rotated_", mu=0.0, sd=1.0, shape=shape, **kwargs)
-            f = pm.Deterministic(name, mu + tt.dot(chol, v))
+            f = pm.Deterministic(name, mu + tt.dot(cholesky(cov), v))
         else:
-            f = pm.MvNormal(name, mu=mu, chol=chol, shape=shape, **kwargs)
+            f = pm.MvNormal(name, mu=mu, cov=cov, shape=shape, **kwargs)
```

Comment:
Doesn't the cholesky decomp of `cov` get computed twice here?  Once on line 111 and again inside MvNormal?  Or is theano smart enough to not do that?

Reply:
There's an `if else` there so those are two separate branches and only either one of them would be computed, if I'm not mistaken?

Reply:
whoops!  you're absolutely right

Reply:
What's the reasoning for delegating cholesky decomp to MvNormal/StudentT?  Now sometimes the cholesky is computed within this file, and sometimes in MvNormal, potentially with different settings

Reply:
I'm trying to minimize the number of ways and places the same thing is done. In the case of line 111 there's currently no other way.
There might be some computational advantage in having the corr/Cholesky/tau matrices operations living in a separate GraphOp, in which case it would best be managed from a single point. Benchmarking that is the next thing to do/find out once *this* works, which shouldn't be long now.

Eventually, ideally, the Cholesky aspect would be abstracted away entirely in model definitions (triangular decomposition etc. should really happen in the background IMHO).

Reply:
I did notice a few remaining inconsistencies tho, now amended. 

And at second thought… the MvNormal implementation (with any of chol, cov, tau) does sanity checks that are entirely superfluous in the context of GP. As, in fact, does theano and, in turn, scipy. So there's some, albeit tiny, overhead.

So in theory there's some argument for a more distinct implementation when the chol/cov are known to be positive definite. In practice though, I'm not sure that it would be worth the effort.

Reply:
"known to be positive definite" can be a bit tricky. It is entirely possible that a matrix is posdef, but the cholesky decomposition still fails, because there can be numerical issues if some eigenvalues are very small.

Reply:
yup, this is the case for GPs, which is why `stabilize` is used.  The covariance functions are posdef, but computing the Cholesky often fails for numerical reasons.